### ISSUE TITLE
build failure: ambiguous overload in chainparams.cpp

### ISSUE THREAD

**Issue by allesklaar** [FeatherCoin/Feathercoin#132]
On Arch Linux (GCC 6.2.1, boost 1.61.0) 0.9.3.2 build throws numerous `ambiguous overload` errors for chainparams.cpp, as below. 0.9.3.1 build has the same problem.

This looks similar to errors recently reported for Litecoin (https://github.com/litecoin-project/litecoin/issues/240, https://github.com/litecoin-project/litecoin/issues/245, https://bugs.gentoo.org/show_bug.cgi?id=594686, https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=811917 ).

```
chainparams.cpp: In constructor ‘CMainParams::CMainParams()’:
chainparams.cpp:85:52: error: ambiguous overload for ‘operator=’ (operand types are ‘std::vector<unsigned char>’ and ‘boost::assign_detail::generic_list<int>’)
         base58Prefixes[PUBKEY_ADDRESS] = list_of(14); // FeatherCoin addresses start with F
                                                    ^
In file included from /usr/include/c++/6.2.1/vector:69:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/vector.tcc:167:5: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
     vector<_Tp, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/vector:64:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/stl_vector.h:450:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(vector&& __x) noexcept(_Alloc_traits::_S_nothrow_move())
       ^~~~~~~~
/usr/include/c++/6.2.1/bits/stl_vector.h:471:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::initializer_list<_Tp>) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(initializer_list<value_type> __l)
       ^~~~~~~~
chainparams.cpp:86:55: error: ambiguous overload for ‘operator=’ (operand types are ‘std::vector<unsigned char>’ and ‘boost::assign_detail::generic_list<int>’)
         base58Prefixes[SCRIPT_ADDRESS_OLD] = list_of(5);
                                                       ^
In file included from /usr/include/c++/6.2.1/vector:69:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/vector.tcc:167:5: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
     vector<_Tp, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/vector:64:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/stl_vector.h:450:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(vector&& __x) noexcept(_Alloc_traits::_S_nothrow_move())
       ^~~~~~~~
/usr/include/c++/6.2.1/bits/stl_vector.h:471:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::initializer_list<_Tp>) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(initializer_list<value_type> __l)
       ^~~~~~~~
chainparams.cpp:88:51: error: ambiguous overload for ‘operator=’ (operand types are ‘std::vector<unsigned char>’ and ‘boost::assign_detail::generic_list<int>’)
         base58Prefixes[SCRIPT_ADDRESS] = list_of(5);
                                                   ^
In file included from /usr/include/c++/6.2.1/vector:69:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/vector.tcc:167:5: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
     vector<_Tp, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/vector:64:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/stl_vector.h:450:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(vector&& __x) noexcept(_Alloc_traits::_S_nothrow_move())
       ^~~~~~~~
/usr/include/c++/6.2.1/bits/stl_vector.h:471:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::initializer_list<_Tp>) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(initializer_list<value_type> __l)
       ^~~~~~~~
chainparams.cpp:89:53: error: ambiguous overload for ‘operator=’ (operand types are ‘std::vector<unsigned char>’ and ‘boost::assign_detail::generic_list<int>’)
         base58Prefixes[SECRET_KEY] =     list_of(142);// 14+128
                                                     ^
In file included from /usr/include/c++/6.2.1/vector:69:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/vector.tcc:167:5: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
     vector<_Tp, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/vector:64:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/stl_vector.h:450:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(vector&& __x) noexcept(_Alloc_traits::_S_nothrow_move())
       ^~~~~~~~
/usr/include/c++/6.2.1/bits/stl_vector.h:471:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::initializer_list<_Tp>) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(initializer_list<value_type> __l)
       ^~~~~~~~
chainparams.cpp:90:72: error: ambiguous overload for ‘operator=’ (operand types are ‘std::vector<unsigned char>’ and ‘boost::assign_detail::generic_list<int>’)
         base58Prefixes[EXT_PUBLIC_KEY] = list_of(0x04)(0x88)(0xBC)(0x26);
                                                                        ^
In file included from /usr/include/c++/6.2.1/vector:69:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/vector.tcc:167:5: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
     vector<_Tp, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/vector:64:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/stl_vector.h:450:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(vector&& __x) noexcept(_Alloc_traits::_S_nothrow_move())
       ^~~~~~~~
/usr/include/c++/6.2.1/bits/stl_vector.h:471:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::initializer_list<_Tp>) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(initializer_list<value_type> __l)
       ^~~~~~~~
chainparams.cpp:91:72: error: ambiguous overload for ‘operator=’ (operand types are ‘std::vector<unsigned char>’ and ‘boost::assign_detail::generic_list<int>’)
         base58Prefixes[EXT_SECRET_KEY] = list_of(0x04)(0x88)(0xDA)(0xEE);
                                                                        ^
In file included from /usr/include/c++/6.2.1/vector:69:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/vector.tcc:167:5: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
     vector<_Tp, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/vector:64:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/stl_vector.h:450:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(vector&& __x) noexcept(_Alloc_traits::_S_nothrow_move())
       ^~~~~~~~
/usr/include/c++/6.2.1/bits/stl_vector.h:471:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::initializer_list<_Tp>) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(initializer_list<value_type> __l)
       ^~~~~~~~
chainparams.cpp: In constructor ‘CTestNetParams::CTestNetParams()’:
chainparams.cpp:157:52: error: ambiguous overload for ‘operator=’ (operand types are ‘std::vector<unsigned char>’ and ‘boost::assign_detail::generic_list<int>’)
         base58Prefixes[PUBKEY_ADDRESS] = list_of(65);
                                                    ^
In file included from /usr/include/c++/6.2.1/vector:69:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/vector.tcc:167:5: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
     vector<_Tp, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/vector:64:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/stl_vector.h:450:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(vector&& __x) noexcept(_Alloc_traits::_S_nothrow_move())
       ^~~~~~~~
/usr/include/c++/6.2.1/bits/stl_vector.h:471:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::initializer_list<_Tp>) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(initializer_list<value_type> __l)
       ^~~~~~~~
chainparams.cpp:158:53: error: ambiguous overload for ‘operator=’ (operand types are ‘std::vector<unsigned char>’ and ‘boost::assign_detail::generic_list<int>’)
         base58Prefixes[SCRIPT_ADDRESS] = list_of(196);
                                                     ^
In file included from /usr/include/c++/6.2.1/vector:69:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/vector.tcc:167:5: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
     vector<_Tp, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/vector:64:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/stl_vector.h:450:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(vector&& __x) noexcept(_Alloc_traits::_S_nothrow_move())
       ^~~~~~~~
/usr/include/c++/6.2.1/bits/stl_vector.h:471:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::initializer_list<_Tp>) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(initializer_list<value_type> __l)
       ^~~~~~~~
chainparams.cpp:159:53: error: ambiguous overload for ‘operator=’ (operand types are ‘std::vector<unsigned char>’ and ‘boost::assign_detail::generic_list<int>’)
         base58Prefixes[SECRET_KEY]     = list_of(193);//65+128
                                                     ^
In file included from /usr/include/c++/6.2.1/vector:69:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/vector.tcc:167:5: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
     vector<_Tp, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/vector:64:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/stl_vector.h:450:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(vector&& __x) noexcept(_Alloc_traits::_S_nothrow_move())
       ^~~~~~~~
/usr/include/c++/6.2.1/bits/stl_vector.h:471:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::initializer_list<_Tp>) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(initializer_list<value_type> __l)
       ^~~~~~~~
chainparams.cpp:160:72: error: ambiguous overload for ‘operator=’ (operand types are ‘std::vector<unsigned char>’ and ‘boost::assign_detail::generic_list<int>’)
         base58Prefixes[EXT_PUBLIC_KEY] = list_of(0x04)(0x35)(0x87)(0xCF);
                                                                        ^
In file included from /usr/include/c++/6.2.1/vector:69:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/vector.tcc:167:5: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
     vector<_Tp, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/vector:64:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/stl_vector.h:450:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(vector&& __x) noexcept(_Alloc_traits::_S_nothrow_move())
       ^~~~~~~~
/usr/include/c++/6.2.1/bits/stl_vector.h:471:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::initializer_list<_Tp>) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(initializer_list<value_type> __l)
       ^~~~~~~~
chainparams.cpp:161:72: error: ambiguous overload for ‘operator=’ (operand types are ‘std::vector<unsigned char>’ and ‘boost::assign_detail::generic_list<int>’)
         base58Prefixes[EXT_SECRET_KEY] = list_of(0x04)(0x35)(0x83)(0x94);
                                                                        ^
In file included from /usr/include/c++/6.2.1/vector:69:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/vector.tcc:167:5: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(const std::vector<_Tp, _Alloc>&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
     vector<_Tp, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/vector:64:0,
                 from serialize.h:21,
                 from bignum.h:9,
                 from chainparams.h:9,
                 from chainparams.cpp:6:
/usr/include/c++/6.2.1/bits/stl_vector.h:450:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(vector&& __x) noexcept(_Alloc_traits::_S_nothrow_move())
       ^~~~~~~~
/usr/include/c++/6.2.1/bits/stl_vector.h:471:7: note: candidate: std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::initializer_list<_Tp>) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]
       operator=(initializer_list<value_type> __l)
       ^~~~~~~~
```

**Comment by aciddude** [comment]
Try going down to boost158 or boost155 and make sure you build boost from source with the same gcc compiler you're using for feathercoin.

**Comment by wellenreiter01** [comment]
@allesklaar 
Did accidude's suggetion solve your problem? 
if it is related to boost, it's hard to fix in the chainparms code

**Comment by allesklaar** [comment]
Downgrading boost is not a solution for a real-life system in the long run.

This problem now probably applies to a limited set of systems, which already have newer versions of boost, but soon will spread over to all other distors once they catch up with boost upstream. The problem needs to be solved in the FTC client code. Sorry I can't help with that. I was hoping there were FTC developers here who would know how to do it.

**Comment by aciddude** [comment]
Hi @allesklaar  
I've managed to recreate and fix your error.  I've notified the Feathercoin Developers and they will either provide a fix to the 0.9.3.1 branch or apply the fix on the next release which is due soon. 

Thank you for flagging this. 

acid.

**Comment by allesklaar** [comment]
@aciddude

Thanks. Your patch was merged into FeatherCoin:0.9.3.1 2 days ago. I still can't build it though - another error crops out:

```
  CXX      coins.o
coins.cpp: In member function ‘double CCoinsViewCache::GetPriority(const CTransaction&, int)’:
coins.cpp:191:18: error: base operand of ‘->’ has non-pointer type ‘const CCoins’
         if (coins->nHeight <= nHeight) {
                  ^~
coins.cpp:192:84: error: base operand of ‘->’ has non-pointer type ‘const CCoins’
             dResult += (double)(coins.vout[txin.prevout.n].nValue) * (nHeight-coins->nHeight);
                                                                                    ^~
```

This is with the current stock Arch Linux build stack: QMake 3.0 using Qt  5.7.1, boost 1.62, gcc 6.2.1. My build command was:

```
./autogen.sh
./configure --with-incompatible-bdb
make
```

**Comment by allesklaar** [comment]
I have reported the coins.cpp issue seperately in #142.

**Comment by wellenreiter01** [comment]
issue #132 will be fixed in 0.9.6 (next release-candidate)
